### DISCUSSION TITLE
How to keep asyncio task running continuously?

### DISCUSSION THREAD

**Discussion by neliel123** [micropython/micropython#13066] [General]
Hi,
I have a coroutine that I wanted to continuously run and it keeps on polling my REST API server.  However, for some reason using urequests fails for some time it works for an hour or so then it stops working and stop sending API call.
I think that memory gets eaten up which is bad as my system is not stable

My app runs on batteries so I do not know what is happening.

Not sure but if I do it like this, will this solve my problem?
```
import asyncio
 
# In a continuous while loop, a REST API is called using urequests
async def call_rest_api():
    while True:
        print('executing the task')
        # block for a moment
        await asyncio.sleep(1)
        print('raise error in task')
        raise
 
# main coroutine
async def main():
    print('main coroutine started')
    
    # wait for the task to complete
    task = None
    while True:
        try:
            # create and schedule the task
            task = asyncio.create_task(call_rest_api())
            await task
        except:
            print("Restarting task as an unknown error occured")

    print('main coroutine done')
 
# start the asyncio program
asyncio.run(main())

```
I have been reading some discussions about issues with urequests and how it is not compatible with asyncio. Is this true and what are my alternatives?

Fairly new to asyncio but I like the benefits of this in my programming.

My REST API Client code looks like this
```
import urequests

class RestAPIError(Exception):
    pass

class RestClient:
    def __init__(self):
        pass
         
    async def process_api(self):
        response = None
        try:
            resource_url = "MY REST API URL"
            response = urequests.post(resource_url)
            return response.json()
        except Exception as e:
            print("An error occurred during HTTP request:", str(e))
            raise RestAPIError
        finally:
            if response:
                response.close()
```

And the way I call this is like this in my main code actually.
```
# In a continuous while loop, a REST API is called using urequests
async def call_rest_api():
    rest_client = RestClient()
    while True:
        print('executing the task')
        # block for a moment, simulate REST API Call
        await rest_client.process_api()

```

Thanks!

**Comment by karfas** [comment]
> My app runs on batteries so I do not know what is happening.

Nothing prevents you from connecting the device to a PC and leave it running overnight. Connect to the REPL and **see** what's happening.

I guess that successful `urequests.post()` responses are eating up memory (there is no `response.close()` call for them).

**Reply by karfas** [reply]
Forget the last sentence of my last comment (finally will always be called in python try/except/finally blocks).

**Comment by peterhinch** [comment]
As a general comment `asyncio` applications can run indefinitely. Keeping WiFi connected applications running indefinitely is another matter. See [this doc](https://github.com/peterhinch/micropython-samples/blob/master/resilient/README.md).
